pkg/networkservice/chains/nsmgr/server_heal_test.go

```diff
@@ -0,0 +1,371 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package nsmgr_test define a tests for NSMGR chain element.
+package nsmgr_test
+
+import (
+	"context"
+	"runtime"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/api/pkg/api/networkservice/mechanisms/cls"
+	kernelmech "github.com/networkservicemesh/api/pkg/api/networkservice/mechanisms/kernel"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/goleak"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/sandbox"
+)
+
+const (
+	tick    = 10 * time.Millisecond
+	timeout = 10 * time.Second
+)
+
+func TestNSMGR_HealEndpoint(t *testing.T) {
+	defer goleak.VerifyNone(t, goleak.IgnoreCurrent())
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+
+	defer cancel()
+	domain := sandbox.NewBuilder(t).
+		SetNodesCount(2).
+		SetRegistryProxySupplier(nil).
+		SetContext(ctx).
+		Build()
+	defer domain.Cleanup()
+
+	expireTime, err := ptypes.TimestampProto(time.Now().Add(time.Second))
+	require.NoError(t, err)
+
+	nseReg := &registry.NetworkServiceEndpoint{
+		Name:                "final-endpoint",
+		NetworkServiceNames: []string{"my-service-remote"},
+		ExpirationTime:      expireTime,
+	}
+
+	counter := &counterServer{}
+	nseCtx, nseCtxCancel := context.WithTimeout(context.Background(), time.Second)
+	defer nseCtxCancel()
+
+	_, err = domain.Nodes[0].NewEndpoint(nseCtx, nseReg, sandbox.GenerateExpiringToken(time.Second), counter)
+	require.NoError(t, err)
+
+	request := &networkservice.NetworkServiceRequest{
+		MechanismPreferences: []*networkservice.Mechanism{
+			{Cls: cls.LOCAL, Type: kernelmech.MECHANISM},
+		},
+		Connection: &networkservice.Connection{
+			Id:             "1",
+			NetworkService: "my-service-remote",
+			Context:        &networkservice.ConnectionContext{},
+		},
+	}
+
+	nsc := domain.Nodes[1].NewClient(ctx, sandbox.GenerateTestToken)
+
+	conn, err := nsc.Request(ctx, request.Clone())
+	require.NoError(t, err)
+	require.NotNil(t, conn)
+	require.Equal(t, 1, counter.UniqueRequests())
+	require.Equal(t, 8, len(conn.Path.PathSegments))
+
+	nseReg2 := &registry.NetworkServiceEndpoint{
+		Name:                "final-endpoint2",
+		NetworkServiceNames: []string{"my-service-remote"},
+	}
+	_, err = domain.Nodes[0].NewEndpoint(ctx, nseReg2, sandbox.GenerateTestToken, counter)
+	require.NoError(t, err)
+
+	nseCtxCancel()
+
+	// Wait NSE expired and reconnecting to the new NSE
+	require.Eventually(t, checkSecondRequestsReceived(counter.UniqueRequests), timeout, tick)
+
```

Comment:
If we add here refresh attempt, it fails:
```go
// Refresh
request = request.Clone()
request.Connection = conn.Clone()

conn, err = nsc.Request(ctx, request.Clone())
require.NoError(t, err)
```
Results in:
```
...
    server_heal_test.go:106: 
        	Error Trace:	server_heal_test.go:106
        	Error:      	Received unexpected error:
        	            	rpc error: code = Unknown desc = Error returned from api/pkg/api/networkservice/networkServiceClient.Request: rpc error: code = Unknown desc = Error returned from api/pkg/api/networkservice/networkServiceClient.Request: rpc error: code = Unknown desc = Error returned from api/pkg/api/networkservice/networkServiceClient.Request: rpc error: code = Unknown desc = Error returned from api/pkg/api/networkservice/networkServiceClient.Request: rpc error: code = Unknown desc = Error returned from api/pkg/api/networkservice/networkServiceClient.Request: rpc error: code = Unknown desc = new selected endpoint URL tcp://127.0.0.1:50786 doesn't match endpoint URL selected before interpose NSE tcp://127.0.0.1:50794
        	            	Error returned from api/pkg/api/networkservice/networkServiceClient.Request
        	            	github.com/networkservicemesh/sdk/pkg/networkservice/core/trace.(*beginTraceClient).Request
        	            		/Users/user/Projects/NSM/sdk/pkg/networkservice/core/trace/client.go:62
...
```

Reply:
Good catch. Fixed